File: src/main/java/aurocosh/divinefavor/common/tasks/base/BaseTask.java
package aurocosh.divinefavor.common.tasks.base;

import net.minecraft.world.World;
import net.minecraftforge.common.MinecraftForge;

public class BaseTask {
    private boolean isRunning;

    protected final World world;
    public BaseTask(World world) {
        this.world = world;
    }

    public boolean isRunning() {
        return isRunning;
    }

    protected boolean isSameDimension(World world) {
        return this.world.provider.getDimension() == world.provider.getDimension();
    }

    public void start() {
        isRunning = true;
        MinecraftForge.EVENT_BUS.register(this);
    }

    public void finish() {
        isRunning = false;
        MinecraftForge.EVENT_BUS.unregister(this);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/tasks/base/BaseTask.kt
package aurocosh.divinefavor.common.tasks.base

import net.minecraft.world.World
import net.minecraftforge.common.MinecraftForge

open class BaseTask(protected val world: World) {
    var isRunning: Boolean = false
        private set

    private val finishActions = ArrayList<(World) -> Unit>()
    fun addFinishAction(action: (World) -> Unit) = finishActions.add(action)

    protected fun isSameDimension(world: World): Boolean {
        return this.world.provider.dimension == world.provider.dimension
    }

    open fun start() {
        isRunning = true
        MinecraftForge.EVENT_BUS.register(this)
        finishActions.forEach { it.invoke(world) }
    }

    fun finish() {
        isRunning = false
        MinecraftForge.EVENT_BUS.unregister(this)
    }
}